doc/resizefs.reiser4.8
.\"						Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH resizefs.reiser4 8 "20 Jul, 2003" reiser4progs "reiser4progs manual"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
resizefs.reiser4 \- the program for resize reiser4 filesystem.
.SH SYNOPSIS
.B resizefs.reiser4
[ options ] FILE size[K|M|G]
.SH DESCRIPTION
.B resizefs.reiser4
is reiser4 filesystem resize program.
.SH COMMON OPTIONS
.TP
.B -V, --version
prints program version.
.TP
.B -?, -h, --help
prints program help.
.TP
.B -y, --yes
assumes an answer 'yes' to all questions.
.TP
.B -f, --force
forces resizefs to use whole disk, not block device or mounted partition.
.TP
.B -c, --cache N
sets tree cache node number to passed value. This affects very much
behavior of libreiser4. It affects speed, tree allocation, etc.
.SH PLUGIN OPTIONS
.TP
.B -p, --print-profile
prints the plugin profile. This is the set of default plugins used for all
parts of a filesystem -- format, nodes, files, directories, hashes, etc. If
--override is specified, then prints modified plugins.
.TP
.B -l, --print-plugins
prints all plugins libreiser4 know about.
.TP
.B -o, --override TYPE=PLUGIN, ...
overrides the default plugin of the type "TYPE" by the plugin "PLUGIN" in 
the plugin profile.
.sp 1
Examples:
.sp 1
resizefs.reiser4 -qf -o nodeptr=nodeptr41,hash=rupasov_hash /dev/hda2 /dev/hda3
.RS
.SH REPORTING BUGS
Report bugs to <umka@namesys.com>, <reiserfs-list@namesys.com>
.SH SUPPORT
You can get support at http://www.namesys.com/support.html
.SH SEE ALSO
.BR measurefs.reiser4(8),
.BR debugfs.reiser4(8),
.BR fsck.reiser4(8)
.BR mkfs.reiser4(8)
.BR copyfs.reiser4(8)
.SH AUTHOR
This manual page was written by Yury Umanets <umka@namesys.com>



ChangeLog
This file is part of reiser4progs package

Copyright (C) 2001, 2002, 2003, 2004 by Hans Reiser, licensing governed by
reiser4progs/COPYING.

This file may be modified and/or distributed without restriction. This is
not an invitation to misrepresent the history of reiser4progs.

2005-08-12
	1.0.5 version.
2005-08-10
	Installation fix.
	Return value fixes.
	Backup block handling fixes.
2005-06-28
	Ability to link against libaal placed at arbitrary location is fixed.
	Unlink dir fix.
	Stat40 check_struct fix.
	Man pages for not ready utils are removed.
	Handle new (crc related) fields in plugin set correctly.
	Flush fs metadata fix (interrupt related).
	Do not flush dirty nodes if fsck fails.
	Recalculate bitmap checksums at the end of fsck if they are wrong.
	Do not update bitmaps after every fsck pass if no metadata found.
2005-03-23
	nomkid option to fsck -- ability to recover reiser4 w/out mkid check.
2005-03-21
	Large file handling fixes.
2005-03-16
	Some casting fixes.
2005-02-21
	Reg40 file body convertion fixes.
2005-02-20
	The 1.0.4 version.
2005-02-19
	A bugfix in file body convertion code.
2005-02-17
	Enable libminimal by default.
	Do not configure empty utils(resizer, cpfs).
2005-01-30
	Some bug fixes in tree balancing, fs check, syncing code.
2005-01-25
	Some bug fixes in tree balancing code.
2005-01-20
	Bug fixes:
	- in backup layout traverse code;
	- in insert extent item-by-item code;
2005-01-17
	A way to recover fs metadata independently of the mkfs id is added.
	Bug fixes:
	- in repair code then a node has a pointer to itself;
	- in not aligned access to different object on-disk data.
2004-12-23
	Bug fixes:
	- in initializing object id fields in SB;
	- in free blocks accounting;
	- in journal unpack code.
2004-12-12
	Some bugfixes:
	- in object/extent truncate code are fixed.
	- in backup handling code is fixed.
	- sware about libaal version correctly.
2004-12-04
	The version is 1.0.3
2004-12-01
	The Pack/unpack journal code is written.
	Cleanups in program options.
2004-11-25
	Build either all or none progs statically.
	The recover from unused space code is added. Could be used as undelete.
2004-11-24
	The backup code is written.
2004-10-29
	A bunch of bugs is fixed.
2004-10-21
	Object detach/unlink methods are fixed.
	Fix the handling of special entries ('.', '..') in directory objects.
	The version is 1.0.2.
2004-10-15
	File body write code (namely extent40 write) is rewritten
	and does work now.
2004-10-12
	Make allocator CRC endian safe
2004-10-07
	A bunch of bugs is fixed in the balancing code
2004-09-22
	Change "stand alone" to "minimal footprint"
2004-09-21
	Get rid of key collision problem is tree_scan
2004-09-16
	sd_bytes fix for files larger then 4G.
2004-09-14
	Make ldconfig to run automatically on 'make install'.
2004-09-13
	Make r5 hash working correctly with extended ASCII characters.
2004-08-24
	Some bugs are fixed:
	- in tree_scan method;
	- in reading extent items;
	- in man pages.
	Some optimisations for stand-alone are added.
	Version is 1.0.1.
2004-08-09
	1.0.0 version is released.


progs/resizefs/resizefs.c
/* Copyright (C) 2001-2005 by Hans Reiser, licensing governed by
   reiser4progs/COPYING.
   
   resizefs.c -- program for resizing reiser4. */

#ifdef HAVE_CONFIG_H
#  include <config.h> 
#endif

#include <stdio.h>
#include <getopt.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <fcntl.h>
#include <sys/stat.h>

#include <misc/misc.h>
#include <reiser4/libreiser4.h>

typedef enum behav_flags {
	BF_FORCE      = 1 << 0,
	BF_YES        = 1 << 1,
	BF_SHOW_PARM  = 1 << 2,
	BF_SHOW_PLUG  = 1 << 3
} behav_flags_t;

/* Prints resizefs options */
static void resizefs_print_usage(char *name) {
	fprintf(stderr, "Usage: %s [ options ] FILE size[K|M|G]\n", name);
    
	fprintf(stderr, 
		"Plugins options:\n"
		"  -p, --print-profile           prints default profile.\n"
		"  -l, --print-plugins           prints known plugins.\n"
	        "  -o, --override TYPE=PLUGIN    overrides the default plugin of the type\n"
	        "                                \"TYPE\" by the plugin \"PLUGIN\" in the\n"
		"                                profile.\n"
		"Common options:\n"
		"  -?, -h, --help                prints program usage.\n"
		"  -V, --version                 prints current version.\n"
		"  -y, --yes                     assumes an answer 'yes' to all questions.\n"
		"  -f, --force                   makes resizer to use whole disk, not\n"
		"                                block device or mounted partition.\n"
		"  -c, --cache N                 number of nodes in tree buffer cache\n");
}

/* Initializes exception streams used by resizefs */
static void resizefs_init(void) {
	int ex;

	/* Setting up exception streams. */
	for (ex = 0; ex < EXCEPTION_TYPE_LAST; ex++)
		misc_exception_set_stream(ex, stderr);
}

int main(int argc, char *argv[]) {
	int c;
	struct stat st;
	char *host_dev;
	count_t fs_len;

	uint32_t cache;
	uint32_t flags = 0;
	char override[4096];

	reiser4_fs_t *fs;
	aal_device_t *device;
	
	static struct option long_options[] = {
		{"version", no_argument, NULL, 'V'},
		{"help", no_argument, NULL, 'h'},
		{"force", no_argument, NULL, 'f'},
		{"yes", no_argument, NULL, 'y'},
		{"print-profile", no_argument, NULL, 'p'},
		{"print-plugins", no_argument, NULL, 'l'},
		{"override", required_argument, NULL, 'o'},
		{"cache", required_argument, NULL, 'c'},
		{0, 0, 0, 0}
	};

	resizefs_init();
	memset(override, 0, sizeof(override));

	/* Parsing parameters */    
	while ((c = getopt_long(argc, argv, "Vhyfo:plc:?",
				long_options, (int *)0)) != EOF) 
	{
		switch (c) {
		case 'h':
		case '?':
			resizefs_print_usage(argv[0]);
			return NO_ERROR;
		case 'V':
			misc_print_banner(argv[0]);
			return NO_ERROR;
		case 'f':
			flags |= BF_FORCE;
			break;
		case 'y':
			flags |= BF_YES;
			break;
		case 'p':
			flags |= BF_SHOW_PARM;
			break;
		case 'l':
			flags |= BF_SHOW_PLUG;
			break;
		case 'c':
			if ((cache = misc_str2long(optarg, 10)) == INVAL_DIG) {
				aal_error("Invalid cache value specified (%s).",
					  optarg);
				return USER_ERROR;
			}

			misc_mpressure_setup(cache);
			break;
		case 'o':
			aal_strncat(override, optarg, aal_strlen(optarg));
			aal_strncat(override, ",", 1);
			break;
		}
	}
	
	if (optind != argc - 1 && optind != argc - 2) {
		resizefs_print_usage(argv[0]);
		goto error;
	}
	
	if (!(flags & BF_YES))
		misc_print_banner(argv[0]);

	if (libreiser4_init()) {
		aal_error("Can't initialize libreiser4.");
		goto error;
	}

	/* Overriding params by passed values. This should be done after
	   libreiser4 is initialized. */
	if (aal_strlen(override) > 0) {
		override[aal_strlen(override) - 1] = '\0';

		if (!(flags & BF_YES)) {
			aal_mess("Overriding the plugin profile by \"%s\".",
				 override);
		}
		
		if (misc_profile_override(override))
			goto error_free_libreiser4;
	}
	
	if (flags & BF_SHOW_PARM)
		misc_profile_print();

	if (flags & BF_SHOW_PLUG)
		misc_plugins_print();

	if (optind >= argc)
		goto error_free_libreiser4;
		
	host_dev = argv[optind];
	optind++;
    
	if (stat(host_dev, &st) == -1) {
		aal_error("Can't stat %s. %s.", host_dev,
			  strerror(errno));
		goto error_free_libreiser4;
	}
	
	/* Checking if passed device is a block one. If so, we check also is it
	   whole drive or just a partition. If the device is not a block device,
	   then we emmit exception and propose user to use -f flag to force. */
	if (!S_ISBLK(st.st_mode)) {
		if (!(flags & BF_FORCE)) {
			aal_error("Device %s is not block device. "
				  "Use -f to force over.", host_dev);
			goto error_free_libreiser4;
		}
	} else {
		if (((IDE_DISK_MAJOR(MAJOR(st.st_rdev)) && MINOR(st.st_rdev) % 64 == 0) ||
		     (SCSI_BLK_MAJOR(MAJOR(st.st_rdev)) && MINOR(st.st_rdev) % 16 == 0)) &&
		    (!(flags & BF_FORCE)))
		{
			aal_error("Device %s is an entire harddrive, not "
				  "just one partition.", host_dev);
			goto error_free_libreiser4;
		}
	}
   
	/* Checking if passed partition is mounted */
	if (misc_dev_mounted(host_dev) > 0 && !(flags & BF_FORCE)) {
		aal_error("Device %s is mounted at the moment. "
			  "Use -f to force over.", host_dev);
		goto error_free_libreiser4;
	}

	/* Opening device with file_ops and default blocksize */
	if (!(device = aal_device_open(&file_ops, host_dev,
				       512, O_RDWR)))
	{
		aal_error("Can't open %s. %s.", host_dev,
			  strerror(errno));
		goto error_free_libreiser4;
	}

	/* Open file system on the device */
	if (!(fs = reiser4_fs_open(device, 1))) {
		aal_error("Can't open reiser4 on %s", host_dev);
		goto error_free_device;
	}

	/* Get wanted fs len in kilos. */
	if (optind == argc - 1) {
		fs_len = misc_size2long(argv[optind]);
	} else {
		fs_len = reiser4_format_len(device, 1024);
	}
	
	if (fs_len == INVAL_DIG) {
		aal_error("Invalid new filesystem "
			  "size %s.", argv[optind]);
		goto error_free_fs;
	}
	
	fs_len /= (reiser4_master_get_blksize(fs->master) / 1024);

	if (reiser4_fs_resize(fs, fs_len)) {
		aal_error("Can't resize reiser4 on %s.", host_dev);
		goto error_free_fs;
	}
	
	/* Deinitializing filesystem instance and device instance */
	reiser4_fs_close(fs);
	aal_device_close(device);
    
	/* Deinitializing libreiser4. At the moment only plugins are unloading
	   during this. */
	libreiser4_fini();
	return NO_ERROR;

 error_free_fs:
	reiser4_fs_close(fs);
 error_free_device:
	aal_device_close(device);
 error_free_libreiser4:
	libreiser4_fini();
 error:
	return OPER_ERROR;
}
